### PR TITLE
Fix Status to State Mapping page invalid Deployment Id error

### LINKED ISSUE #6007: 500 Server Error For Invalid Deployment Ids on Recurring Donations - Status to State Mapping Page
Customers may see a “500 Server Error” when navigating to NPSP Settings → Recurring Donations → Status to State Mapping.
Screen Shot 2020-11-06 at 2 11 47 PM

### Root Cause
We check for a deployment Id in the  Status Mapping Deployment Id field of the Recurring Donation Custom Setting and then attempt to retrieve information about it if set. (It’s set if you’ve added any custom mappings). If the deployment Id is invalid, a JSON Parsing error occurs because no data is actually returned. A deployment ID is invalid in 2 scenarios:

1. It’s older than 30 days. 
2. A sandbox is created or refreshed from a Production org that has a Enhanced Recurring Donations enabled and has at least 1 custom mapping. 

**Relevant Debug Log Info**
```
09:17:15.0 (92572120)|VARIABLE_SCOPE_BEGIN|[4]|apextype|java:common.apex.runtime.ApexObjectType|true|false
09:17:15.0 (92674632)|VARIABLE_ASSIGNMENT|[4]|apextype|"npsp.CMT_MetadataAPI.DeploymentResponse"|0x13992e87
09:17:15.0 (92739086)|HEAP_ALLOCATE|[143]|Bytes:8
09:17:15.0 (92787952)|HEAP_ALLOCATE|[143]|Bytes:24
09:17:15.0 (92803824)|SYSTEM_METHOD_ENTRY|[1]|JSON.JSON()
09:17:15.0 (92808643)|STATEMENT_EXECUTE|[1]
09:17:15.0 (92814625)|SYSTEM_METHOD_EXIT|[1]|JSON
09:17:15.0 (92832410)|METHOD_ENTRY|[143]||System.JSON.deserialize(String, System.Type)
09:17:15.0 (93263202)|METHOD_EXIT|[143]||System.JSON.deserialize(String, System.Type)
09:17:15.0 (93277703)|METHOD_EXIT|[408]|01p4x000001CqGl|npsp.CMT_MetadataAPI.getDeploymentResult(Id)
09:17:15.0 (93293026)|SYSTEM_METHOD_EXIT|[444]|npsp.RD2_StatusMappingSettings_CTRL.DeploymentHandler.__sfdc_deployResult()
09:17:15.0 (93301708)|METHOD_EXIT|[295]|01p4x000001CqAb|npsp.RD2_StatusMappingSettings_CTRL.DeploymentHandler.hasDeployResult()
09:17:15.0 (93315870)|METHOD_EXIT|[281]|01p4x000001CqAb|npsp.RD2_StatusMappingSettings_CTRL.getDeployResult(String, npsp.RD2_StatusMappingSettings_CTRL.MetadataDeploymentHandler)
09:17:15.0 (93324973)|SYSTEM_MODE_EXIT|false
09:17:15.0 (93449197)|FATAL_ERROR|System.JSONException: Malformed JSON: Expected '{' at the beginning of object
Class.System.JSON.deserialize: line 15, column 1
Class.npsp.CMT_MetadataAPI.getDeploymentResult: line 143, column 1
Class.npsp.RD2_StatusMappingSettings_CTRL.DeploymentHandler.__sfdc_deployResult: line 408, column 1
Class.npsp.RD2_StatusMappingSettings_CTRL.DeploymentHandler.hasDeployResult: line 444, column 1
Class.npsp.RD2_StatusMappingSettings_CTRL.getDeployResult: line 295, column 1
Class.npsp.RD2_StatusMappingSettings_CTRL.getDeployResult: line 281, column 1
09:17:15.0 (93468499)|FATAL_ERROR|System.JSONException: Malformed JSON: Expected '{' at the beginning of object

```

### Workaround

- Go to Setup → Custom Settings.
- Click Manage next to the Recurring Donations Settings and click Edit.
- Clear out the value in the “Status Mapping Deployment Id” field.

### Steps to Repeat

1. Enable Enhanced Recurring Donations
2. Add a new value to the Recurring Donation npsp__Status__c field.
3. Go to  NPSP Settings → Recurring Donations → Status to State. Click Edit. 
4. Select a State for your new Status and hit Save.
5. Go to Setup → Custom Settings → and click Manage next to the Recurring Donations Settings.
6. Note the Deployment Id listed in the Status Mapping Deployment Id field. (should look something like: 0Af01000005rGtmCAE). 
7. Click Edit and change the deployment Id to something else entirely. For example, change the last 3 letters. (This is to simulate a deployment Id that is not present in the org within the last 30 days). 
8. Go back to NPSP Settings → Recurring Donations → Status to State.
9. Confirm the 500 Server Error appears and no mapped statuses are displayed.

### Potential Solution
We had a similar issue occur with Advanced Mapping. The PR that resolved the issue is here:
https://github.com/SalesforceFoundation/NPSP/pull/4700 (https://github.com/SalesforceFoundation/NPSP/pull/4700)

### PR DESCRIPTION
**Pull request by npsp-reedestockton**
# Critical Changes

# Changes
 - We fixed an issue that could cause the Recurring Donations Status to State Mapping page to produce a 500 Server error.

# Issues Closed
Fixes #6007

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by npsp-reedestockton** [comment]
Tracking [W-8350830](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008KqoYIAS/view)

**Comment by npsp-reedestockton** [comment]
@andrewyu-salesforce @RadSFDeveloper Ready for your review.
@sam-knox Would you please review the PR change description when you have a moment?

Thanks all!

File: src/classes/RD2_StatusMappingSettings_CTRL.cls [hunk 1]
```diff
@@ -405,7 +407,15 @@ public with sharing class RD2_StatusMappingSettings_CTRL {
             get {
                 if (deployResult == null) {
                     if (String.isNotBlank(deploymentId)) {
-                        deployResult = CMT_MetadataAPI.getDeploymentResult(deploymentId);
+                        try {
+                            deployResult = CMT_MetadataAPI.getDeploymentResult(deploymentId);
+                        }
+                        catch (Exception e) {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_StatusMappingSettings_CTRL.cls:413]: Let's explicitly set `deployResult = null` in the `catch` block.

**Review by RadSFDeveloper — COMMENTED** [review]
There is one minor comment for deployment result on an error.

I thought it would be a good chance to fix the error construct in this PR, however, it might be better to log a new WI to fix the `handleError()` error message construct to display the error content. For example, if the `try/catch` block is removed, and the invalid Id is passed in, then the error message on the page should display the error detail instead of "Server error 500":
Screen Shot 2021-01-06 at 9 44 06 AM

**Review by RadSFDeveloper — APPROVED** [review]
Looks good, thanks @npsp-reedestockton !

**Review by sam-knox — APPROVED** [review]

**Comment by rponti-sforg** [comment]
QE testing is now complete for this PR.  Based on the WI Description, ACs, and PR comments, tests have passed as expected.

For further testing details:
Test Charter: https://salesforce.quip.com/flXHAIitYa2b
WI: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008KqoYIAS/view

Corresponding WI status has been set to 'Closed.

**Pull request merged**